### PR TITLE
Add pbgcpp pbmm2

### LINKED ISSUE #4443: Add tool GenomicConsensus (Arrow)
Hi,

I'm working on a wrapper for PacBio's [GenomicConsensus Arrow](https://github.com/PacificBiosciences/GenomicConsensus) polishing algorithm. I plan to submit the wrapper in a PR once it's ready.

Tom

### PR DESCRIPTION
**Pull request by TomHarrop**
Requesting to add `pbgcpp` (also known as the GenomicConsensus arrow algorithm; fixes #4443) and a tool required to generate the input data, `pbmm2`.

---
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/pbgcpp/pbgcpp.xml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    <command detect_errors="exit_code"><![CDATA[
+        ## set up files
+        #if $reference_source.reference_source_selector == 'history':
+            #set ref_fn = 'reference.fa'
+            ln -f -s '$reference_source.ref_file' '$ref_fn' &&
+        #else:
+            #set ref_fn = $reference_source.ref_file.fields.path
+        #end if
+
+        #set input_fn = 'input.' + $input.ext
```
Comment (bgruening) [inline tools/pbgcpp/pbgcpp.xml:17]: its always BAM, isn't it? So we could simplify it a bit.

File: tools/pbgcpp/pbgcpp.xml [hunk 2]
```diff
@@ -0,0 +1,143 @@
+<tool id="pbgcpp" name="pbgcpp (GenomicConsensus) arrow." version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
```
Comment (bernt-matthias) [inline tools/pbgcpp/pbgcpp.xml:1]: Maybe just 

```suggestion
<tool id="pbgcpp" name="pbgcpp" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
```

File: tools/pbmm2/pbmm2.xml [hunk 3]
```diff
@@ -0,0 +1,102 @@
+<tool id="pbmm2" name="pbmm2:" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
```
Comment (bernt-matthias) [inline tools/pbmm2/pbmm2.xml:1]: Maybe

```suggestion
<tool id="pbmm2" name="pbmm2" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="@PROFILE@">
```

**Review by bernt-matthias — COMMENTED** [review]
Would it make sense to have all tools in one directory and set them up as tool suite?

**Comment by TomHarrop** [comment]
> Would it make sense to have all tools in one directory and set them up as tool suite?

@bernt-matthias Happy to do that if it's preferred, just let me know.

I chose to separate them because these two tools aren't necessarily related or used together, they're released separately and have different dependencies.

On the other hand they are both part of the [PacBio Secondary Analysis Tools on Bioconda](https://github.com/PacificBiosciences/pbbioconda#packages) suite. We could make that a Galaxy suite. However the list of packages contains tools with a wide range of uses, e.g. transcript assembly, amplicon seq etc.

File: tools/pbgcpp/pbgcpp.xml (hunk 1 above)
Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml]: Good point, fixed with https://github.com/galaxyproject/tools-iuc/pull/5125/commits/bfb031a2a55a4f4dfd5c3f52d79213659001de38

File: tools/pbgcpp/pbgcpp.xml [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                <param name="ref_file" type="select" label="Using reference genome" help="Select genome from the list">
+                    <options from_data_table="all_fasta">
+                        <filter type="sort_by" column="2" />
+                        <validator type="no_options" message="No reference genomes are available" />
+                    </options>
+                    <validator type="no_options" message="A built-in reference genome is not available for the build associated with the selected input file"/>
+                </param>
+            </when>
+            <when value="history">
+                <param name="ref_file" type="data" format="fasta,fastq" label="Use the following dataset as the reference sequence" help="You can upload a FASTA or FASTQ sequence to the history and use it as reference" />
```
[the remark this answers is not in the document]
Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml]: Thanks, removed fastq with https://github.com/galaxyproject/tools-iuc/pull/5125/commits/e1a70f47c3bf81101ff0de533060fffa315cb3f8.

I think the fasta needs to be unzipped so I'm happy to let Galaxy do it.

File: tools/pbgcpp/pbgcpp.xml [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            </when>
+        </conditional>
+        <param type="data" name="input" format="bam" label="bam" help="The input BAM alignment file" />
```
[the remark this answers is not in the document]
Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml:50]: Yes, pbgcpp needs sorted bam with an index for variant calling.

File: tools/pbgcpp/pbgcpp.xml (hunk 2 above)
Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml]: fixed

**Review by bgruening — COMMENTED** [review]
Thats great, ready to merge, but can you please add a test for the loc file - for the "cached" case.

**Comment by TomHarrop** [comment]
> Thats great, ready to merge, but can you please add a test for the loc file - for the "cached" case.

Thanks @bgruening , I didn't know you could test that with planemo! Added in https://github.com/galaxyproject/tools-iuc/pull/5125/commits/bfe7be1c67b44c622eb03ed11f1e9e0427eddd92

File: tools/pbgcpp/pbgcpp.xml [hunk 6]
```diff
@@ -134,7 +142,9 @@ pbgcpp expects metadata in the bamfile that most aligners (like
 minimap2) don’t include by default. Align the PacBio reads file using
 `pbmm2 <root?tool_id=pbmm2>`__.
 
+--------------
 
+This tool was wrapped by the Galaxy Australia team.
```
Comment (bgruening) [inline tools/pbgcpp/pbgcpp.xml:147]: Please use https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-creator-organization for such things.

Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml]: @mthang @neoformit what do you think? Looks good to me. Galaxy renders it like this.

screenshot-localhost_9090-2023 02 27-10_59_47

Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml]: That's fixed with https://github.com/galaxyproject/tools-iuc/pull/5125/commits/8ead0ecac7dd9f6389546d5d4ea056e1084a3f82 and https://github.com/galaxyproject/tools-iuc/pull/5125/commits/b0a72da8b86d8457c964f51ee2e277c30d3b7457, so I think we're good to go now.

Thanks for your patience with this less-than-perfect PR, and apologies for making changes after you'd already reviewed it once.

File: tools/pbgcpp/macros.xml [hunk 7]
```diff
@@ -12,4 +12,9 @@
             <xref type="bio.tools">genomicconsensus</xref>
         </xrefs>
     </xml>
+    <xml name="creator">
+        <creator>
+            <organization name="Galaxy Australia" url="https://github.com/usegalaxy-au/tools-au"/>
```
Comment (bgruening) [inline tools/pbgcpp/macros.xml]: Should this be https://www.biocommons.org.au/galaxy-australia or https://github.com/usegalaxy-au ?

File: tools/pbgcpp/pbgcpp.xml (hunk 6 above)
Comment (bgruening) [reply tools/pbgcpp/pbgcpp.xml]: All good, thanks for your patience with us here :)
In the end this is the way we learn :)

File: tools/pbgcpp/.shed.yml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+---
+categories:
+  - Assembly
```
Comment (bernt-matthias) [inline tools/pbgcpp/.shed.yml:3]: Maybe [Variant Analysis](https://toolshed.g2.bx.psu.edu/repository/browse_categories?sort=name&operation=repositories_by_category&id=f9cad7b01a472135) or [Next Gen Mappers](https://toolshed.g2.bx.psu.edu/repository/browse_categories?sort=name&operation=repositories_by_category&id=c9715d71689bf781) fit better?

File: tools/pbgcpp/pbgcpp.xml [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    </tests>
+        <help><![CDATA[
+**What it does**
+
+Compute genomic consensus and call variants relative to the reference.
+
+This tool requires a PacBio BAM file.
+
+You can create one by mapping PacBio reads to the reference genome with
+the `pbmm2 <root?tool_id=pbmm2>`__ tool. When doing this, you have to
```
Comment (bernt-matthias) [inline tools/pbgcpp/pbgcpp.xml:114]: That's a cool way to crosslink between tools (if we can assume that the other tool is installed). Maybe we should document this?

Also interesting to see that the URL linter does not complain...

File: tools/pbgcpp/pbgcpp.xml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+information.
+
+**Input**: Aligned subreads in PacBio BAM format (.bam). Compatible with PacBio Sequel data and RS data with the P6-C4 chemistry.
+
+**Output**: Polished contigs in .fasta format.
+
+**Why am I getting “Missing valid chemistry from input file, is this a
+proper PBBAM input file?”**
+
+pbgcpp expects metadata in the bamfile that most aligners (like
```
Comment (bernt-matthias) [inline tools/pbgcpp/pbgcpp.xml:141]: Can we check for this metadata in the input, e.g. by using a metadata validator?

Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml:141]: I'm not sure how to do that. Can you point me to an example? I can't work it out from the `tool > inputs > param > validator` documentation.

"Metadata" here refers to information in the BAM file itself, not the Galaxy metadata.  Do I need the `load_contents` tag for the dataset or something?

We would have to check the `PL` tag (should be PACBIO) in the BAM header and the `pw` tag for bases, according to the [PBBAM spec](https://github.com/PacificBiosciences/PacBioFileFormats/blob/3.0/BAM.rst).

I could also check it in the command section but the tool fails quickly with an error message if it's the wrong type of BAM.

Comment (bernt-matthias) [reply tools/pbgcpp/pbgcpp.xml:141]: Right, seems not to be possible. 

Just for your information, it should be possible to access a dict: `$input.metadata.bam_header` in the command block. But I would not argue for adding a test .. 

You can leave it as it is.

File: tools/pbgcpp/macros.xml (hunk 7 above)
Comment (TomHarrop) [reply tools/pbgcpp/macros.xml]: It's changed to [site.usegalaxy.org.au](https://site.usegalaxy.org.au) in https://github.com/galaxyproject/tools-iuc/pull/5125/commits/b0a72da8b86d8457c964f51ee2e277c30d3b7457

File: tools/pbgcpp/.shed.yml (hunk 8 above)
Comment (TomHarrop) [reply tools/pbgcpp/.shed.yml]: Thanks, that was a mistake. Fixed in https://github.com/galaxyproject/tools-iuc/pull/5125/commits/4637029c433d05691e495234a895464530827a8b

File: tools/pbgcpp/pbgcpp.xml (hunk 9 above)
Comment (TomHarrop) [reply tools/pbgcpp/pbgcpp.xml:114]: You're right, it only makes sense to do that if the tools get installed together (e.g. a suite). I've removed the relative links in https://github.com/galaxyproject/tools-iuc/pull/5125/commits/7999bc8f385f34a3a993cd886b00983f242269db.

Comment (bgruening) [reply tools/pbgcpp/pbgcpp.xml:114]: @TomHarrop I still the reference here? Is that correct?

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Thanks @TomHarrop

**Comment by bernt-matthias** [comment]
```
Unexpected HTTP status code: 500: {"err_msg": "Metadata may have been defined for some items in revision 'a6d93d0d5328'.  Correct the following problems if necessary and reset metadata.<br/><b>pbgcpp.xml<\/b> - This file requires an entry in the tool_data_table_conf.xml file.  Upload a file named tool_data_table_conf.xml.sample to the repository that includes the required entry to correct this error.<br/><br/>"}
Unexpected HTTP status code: 400: {"err_msg": "You already own a repository named <b>pbgcpp<\/b>, please choose a different name.", "err_code": 400008}
```

**Comment by TomHarrop** [comment]
@bernt-matthias I'm not quite sure if I need to do anything here, sorry.

I tried to add a `.sample` file (copied from the minimap2 recipe) here: https://github.com/galaxyproject/tools-iuc/pull/5172

Please let me know if something else is needed to get this to deploy.

**Comment by bernt-matthias** [comment]
Seems to be fine at the TS: https://toolshed.g2.bx.psu.edu/view/iuc/pbgcpp/a6d93d0d5328

**Comment by TomHarrop** [comment]
I get "Repository does not contain any installable revisions." when i try to install pbgcpp on my VM (Galaxy 22.05).

I can't see pbmm2 on the toolshed at all.

**Pull request merged**